source/agora/network/NodeLocator.d

```diff
@@ -0,0 +1,268 @@
+    {
+        this.log = Logger(__MODULE__);
+        this.locator_db_path = locator_db_path;
+    }
+
+    /// Start the node locator, returns true on succes, false otherwise
+    public bool start ()
+    {
+        import std.string : toStringz;
+
+        int status = MMDB_open(this.locator_db_path.toStringz(), MMDB_MODE_MMAP, &this.mmdb);
+        if (MMDB_SUCCESS != status)
+        {
+            log.error("Can't open {} - {}", this.locator_db_path, MMDB_strerror(status));
+
+            if (MMDB_IO_ERROR == status)
+                log.error("IO error: {}", strerror(errno));
+
+            return false;
+        }
+        return true;
+    }
+
+    /// Stop the node locator
+    public void stop ()
+    {
+        MMDB_close(&this.mmdb);
+        this.mmdb = MMDB_s.init;
+    }
+
+    ///
+    public ~this ()
+    {
+        if (this.mmdb != MMDB_s.init)
+        {
+            fprintf(stderr, "NodeLocator.shutdown() was not called manually");
+            this.stop();
+        }
+    }
+
+    /***************************************************************************
+
+        Extracts certain properties of the geographical location of the node.
+        The properties can be continent, country, ...
+        The exact properties that can be extracted and the query path that
+        identifies the properties are implementation dependent.
+
+        An example query path can be:
+            continent->names->en
+            location->latitude
+
+        Params:
+            address = the network address of the node
+            paths = query parts to identify which properties needs to be extracted
+
+        Returns:
+            an array of extracted properties, if a particular property
+            cannot be extracted, then `MissingValue` is returned
+
+    ***************************************************************************/
+
+    public string[] extractValues (Address address, string[] paths)
+    {
+        import std.algorithm : map;
+        import std.array : array, split;
+        import std.range : repeat;
+        import std.string : toStringz;
+
+        string[] extractedValues;
+        extractedValues.length = paths.length;
+
+        // Look up the value in the database
+        int gai_error;
+        int mmdb_error;
+        MMDB_lookup_result_s result = MMDB_lookup_string(&this.mmdb, address.ptr, &gai_error, &mmdb_error);
+        if (0 != gai_error)
+        {
+            log.warn("Error from getaddrinfo for {}", address);
+            return MissingValue.repeat(paths.length).array();
+        }
+        if (MMDB_SUCCESS != mmdb_error)
+        {
+            log.warn("Got an error from libmaxminddb: {}", MMDB_strerror(mmdb_error));
+            return MissingValue.repeat(paths.length).array();
+        }
+
+        // Extract the data
+        foreach (ind, path; paths)
+        {
+            auto path_splitted = path.split(PathSeparator).map!(path_part => path_part.toStringz()).array();
```

Comment:
IIUC the MMDB API requires an array of `char*` so why do we have custom path logic using `->` ?